File: 113_Path_Sum_II/113_1.cpp
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode *root, int sum) {
        vector<int> v;
        dfs(root, v, sum);
        return res;
    }

    void dfs(TreeNode *root, vector<int> &v, int sum) {
        if (!root) return;
        v.emplace_back(root->val);
        if (!root->left && !root->right && root->val == sum)
            res.emplace_back(v);
        dfs(root->left, v, sum - root->val);
        dfs(root->right, v, sum - root->val);
        v.pop_back();
    }

private:
    vector<vector<int>> res;
};

File: 113_Path_Sum_II/113_1.scala
object Solution {
  def pathSum(root: TreeNode, targetSum: Int): List[List[Int]] = {
    val res = new scala.collection.mutable.ListBuffer[List[Int]]
    def dfs(node: TreeNode, sum: Int, tmp: List[Int]): Unit = {
      if (node != null) {
        val newList = node.value :: tmp
        if (node.left == null && node.right == null && node.value == sum) {
          res += newList.reverse
        } else {
          val newSum = sum - node.value
          dfs(node.left, newSum, newList)
          dfs(node.right, newSum, newList)
        }
      }
    }
    dfs(root, targetSum, Nil)
    res.toList
  }
}
